packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+
+```TS
+public TableColumnText extends FoundationElement implements ITableColumn {
+    ...
+
+    @attr
+    public valueKey: string;
+
+    public getDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public cellTemplateFn(): ViewTemplate<ITableCellData> {
+        return html<ITableCellData>`
+            <nimble-text-field readonly="true" value=${x => x.data[this.valueKey]}>
+            </nimble-text-field>
+        `;
+    }
+}
+```
+
+Enabling a particular column to access multiple pieces of data to use in its rendering is equally trivial:
+
+```TS
+public TableColumnNumberWithUnit extends FoundationElement implements ITableColumn {
+    ...
+
+    @attr
+    public valueKey: string;
+
+    @attr
+    public unitKey: string;
+
+    public getDataKeys(): string[] {
+        return [valueKey, unitKey];
+    }
+
+    public cellTemplateFn(): ViewTemplate<ITableCellData> {
+        return html<ITableCellData>`
+            <nimble-text-field readonly="true" value=${x => this.formatData(x.data[this.valueKey], x.data[this.unitKey])}>
+            </nimble-text-field>
+        `;
+    }
+
+    public formatData(value: unknown, unit: unknown): string {
+        return `${(value as string)}, ${(unit as string)}`;
+    }
+}
+```
+
+Here we have a column visualize in different ways based on custom logic:
+
+```TS
+public TableColumnPositiveNegativeNumber extends FoundationElement implements ITableColumn {
+    ...
+
+    @attr
+    public valueKey: string;
+
+    public getDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public cellTemplateFn(): ViewTemplate<ITableCellData> {
+        return html<ITableCellData>`
+            ${when(x => !this.isPositive(x.data[this.valueKey])),
+            html`
+                <nimble-text-field readonly="true" value=${x => x.data[this.valueKey]} style="fontColor: red;">
+                </nimble-text-field>
+            `}
+            ${when(x => this.isPositive(x.data[this.valueKey])),
+            html`
+                <nimble-text-field readonly="true" value=${x => x.data[this.valueKey]} style="fontColor: green;">
+                </nimble-text-field>
+            `}
+        `;
+    }
+
+    public isPositive(value: unknown): bool {
+        return (value as number) >= 0;
+    }
+}
+```
+
+Finally, here is a column that allows a user to register a callback for a click event on a button inside the cell template:
+
+```TS
+public TableColumnButton extends FoundationElement implements ITableColumn {
+    ...
+
+    @attr
+    public idKey: string;
+
+    public getDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public callback: (id: string) => void;
+
+    public cellTemplateFn(): ViewTemplate<ITableCellData> {
+        return html<ITableCellData>`
+                <nimble-button readonly="true" @click="${x => this.callback(x.data[this.idKey])}>
```

Comment:
I don't think this conceptually makes sense. The context of `this` is the configuration custom element that implements `ITableColumn` the configuration custom element should not be hosting any state / methods, etc used by the cell instances. It's just a dumb configuration object.

That should either just be a function call:

```
const callback = () => {};

public TableColumnButton extends FoundationElement implements ITableColumn {
...
    public cellTemplateFn(): ViewTemplate<ITableCellData> {
        return html<ITableCellData>`
                <nimble-button readonly="true" @click="${x => callback(x.data[this.idKey])}>
```

Or imported from a utilities file. Whatever we do it shouldn't be bound to the configuration custom element itself.

Reply:
Updated to use the event pattern discussed off-line.